api4/user.go

```diff
@@ -532,6 +535,51 @@ func revokeSession(c *Context, w http.ResponseWriter, r *http.Request) {
 	ReturnStatusOK(w)
 }
 
+func autocompleteUsers(c *Context, w http.ResponseWriter, r *http.Request) {
+	channelId := r.URL.Query().Get("in_channel")
+	teamId := r.URL.Query().Get("in_team")
+	term := r.URL.Query().Get("name")
+
+	var autocomplete []*model.User
+	var err *model.AppError
+
+	searchOptions := map[string]bool{}
+
+	hideFullName := !utils.Cfg.PrivacySettings.ShowFullName
+	if hideFullName && !app.SessionHasPermissionTo(c.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		searchOptions[store.USER_SEARCH_OPTION_NAMES_ONLY_NO_FULL_NAME] = true
+	} else {
+		searchOptions[store.USER_SEARCH_OPTION_NAMES_ONLY] = true
+	}
+
+	if len(teamId) > 0 {
+		if !app.SessionHasPermissionToTeam(c.Session, teamId, model.PERMISSION_VIEW_TEAM) {
+			c.SetPermissionError(model.PERMISSION_VIEW_TEAM)
+			return
+		}
+
+		autocomplete, err = app.SearchUsersInTeam(teamId, term, searchOptions)
+	} else if len(channelId) > 0 {
+		if !app.SessionHasPermissionToChannel(c.Session, channelId, model.PERMISSION_READ_CHANNEL) {
+			c.SetPermissionError(model.PERMISSION_READ_CHANNEL)
+			return
+		}
+
+		autocomplete, err = app.SearchUsersInChannel(channelId, term, searchOptions)
+	} else {
+		// No permission check required
+		autocomplete, err = app.SearchUsersInTeam("", term, searchOptions)
+	}
+
+	if err != nil {
+		c.Err = err
+		return
+	} else {
+		w.Write([]byte(model.UserListToJson(autocomplete)))
```

Comment:
The return type is going to be a bit different here than in APIv3. We need to modify it so that it's a consistent JSON object. In APIv3 we had 3 different models for returning autocomplete because we had 3 different endpoints. In APIv4 we have the single endpoint so we should only have the single model, which we need to create called `UserAutocomplete`. It needs to have a field for the actual results (the main list of users) and then the extra fields that get added in the other cases (in team, not in channel, etc.)

This is going to be a little more involved, let me know if you need help